services/QuillLMS/client/app/bundles/PremiumHub/containers/diagnosticGrowthReports/overviewSection.tsx

```diff
@@ -90,33 +90,51 @@ export const OverviewSection = ({
   selectedTimeframe,
   pusherChannel,
   hasAdjustedFiltersFromDefault,
-  handleSetNoDiagnosticDataAvailable
+  hasAdjustedFiltersSinceLastSubmission,
+  handleSetNoDiagnosticDataAvailable,
+  handleTabChangeFromDataChip,
+  handleSetSelectedDiagnosticId,
+  passedData
 }) => {
 
   const [groupByValue, setGroupByValue] = React.useState<DropdownObjectInterface>(groupByDropdownOptions[0])
-  const [loading, setLoading] = React.useState<boolean>(true);
+  const [loading, setLoading] = React.useState<boolean>(!passedData);
   const [preDiagnosticAssignedData, setPreDiagnosticAssignedData] = React.useState<any>(null);
   const [postDiagnosticAssignedData, setPostDiagnosticAssignedData] = React.useState<any>(null);
   const [preDiagnosticCompletedData, setPreDiagnosticCompletedData] = React.useState<any>(null);
   const [postDiagnosticCompletedData, setPostDiagnosticCompletedData] = React.useState<any>(null);
   const [recommendationsData, setRecommendationsData] = React.useState<any>(null);
-  const [aggregatedData, setAggregatedData] = React.useState<any>([]);
+  const [aggregatedData, setAggregatedData] = React.useState<any>(passedData || []);
   const [pusherMessage, setPusherMessage] = React.useState<string>(null)
+  const [updatedFromPusher, setUpdatedFromPusher] = React.useState<boolean>(false)
 
   React.useEffect(() => {
     initializePusher()
   }, [pusherChannel])
 
   React.useEffect(() => {
-    getData()
+    // this is for testing purposes; this value will always be null in a non-testing environment
+    if(!passedData) {
+      getData()
+    }
   }, [searchCount, groupByValue])
 
   React.useEffect(() => {
     if (!pusherMessage) return
 
-    if (filtersMatchHash(pusherMessage)) getData()
+    if (filtersMatchHash(pusherMessage) && !updatedFromPusher) {
+      // this prevents getData from being called multiple times since we receive Pusher messages for each of the 5 different API calls
+      setUpdatedFromPusher(true)
+      getData()
+    }
```

Comment:
I'm confused about how this works -- what if one of the queries is ready way before the other ones? Won't this always just fire when the first query comes back, potentially leaving the user in a state where some of the page has no data?

Reply:
I tested this with multiple different filter changes and it seems to be working as expected. We only really need to worry about matching on one of the five hashes returned to know that we need to refetch data for the newly applied filters. The queries run in parallel and need to be run all together regardless of any change, so we only need to worry about running `getData` once and setting this value to prevent unnecessary subsequent calls.

Further down, we do a check for change in value for `hasAdjustedFiltersSinceLastSubmission` to set `updatedFromPusher` back to false if the filters have been adjusted since the last submission, which will then enable calling `getData` again. 

Does that make sense? We can chat more on Slack if you'd like, I know this is a bit complicated to follow.

Reply:
I'm curious about @anathomical's take here, but I'm still concerned that this implementation is going to lead, maybe very infrequently, to a case where `getData` is called for the second time before all of queries are actually finished running. I think it would be better to wait until the last, not the first of them, has its `pusherMessage` fired (something I think you'd have to do by tracking each of them individually in the state, like `preDiagnosticAssignedQueryFinished`, `preDiagnosticCompletedQueryFinished`, etc, and don't call `getData` again until all five are `true`). I agree that the change would be more complicated code, but I think that's probably worth the trade-off for preventing that edge case. Maybe they all run so fast, or at such a similar speed, that we can be reasonably confident that case will never occur in production, but then we should at least leave a comment to here to that effect because the code itself does not ensure that won't happen.

Reply:
@emilia-friedberg I have an idea. In each of the 5 functions that call a query, I can initially reset their respective pieces of state back to null, like this:

```
  function getPreDiagnosticAssignedData () {
    **setPreDiagnosticAssignedData(null)**
    const searchParams = getSearchParams(PRE_DIAGNOSTIC_ASSIGNED_QUERYSTRING)

    requestPost('/admin_diagnostic_reports/report', searchParams, (body) => {
      if (!body.hasOwnProperty('results')) {
        return
      } else {
        const { results, } = body
        setPreDiagnosticAssignedData(results)
      }
    })
  }
  ```
  
which will ensure that this useEffect accounts for this after initial load (for each query, the result value will always be either an array of values or just an empty array):

```
  React.useEffect(() => {
    if (preDiagnosticAssignedData && postDiagnosticAssignedData && preDiagnosticCompletedData && postDiagnosticCompletedData && recommendationsData) {
      aggregateOverviewData({
        preDiagnosticAssignedData,
        postDiagnosticAssignedData,
        preDiagnosticCompletedData,
        postDiagnosticCompletedData,
        recommendationsData,
        setAggregatedData,
        hasAdjustedFiltersFromDefault,
        handleSetNoDiagnosticDataAvailable,
        setLoading
      })
    }
  }, [preDiagnosticAssignedData, postDiagnosticAssignedData, preDiagnosticCompletedData, postDiagnosticCompletedData, recommendationsData])
  ```

Reply:
This is a good flag from @emilia-friedberg .  I think that the real risk is that there's some hidden complexity on the back-end.  When you make a request for data to an endpoint on the back-end, one of two things happen: 1) the data is in the cache, and you get it back or 2) the data is not in the cache which starts a Sidekiq job to put the data into the cache and then send a Pusher message to let the front-end know that the data is ready.  Consider this scenario:

1. Iniital page load: `getData` triggers 5 separate API calls.  Assume that the cache is empty, so all 5 queries kick off Sidekiq jobs.  Assume that the jobs take 1, 2, 3, 4, and 5 seconds to complete respectively.
2. The first job completes at 1 second, and sends a Pusher messages.  `getData` gets called, making 5 new API calls.  The first query, the one that sent the Pusher notification, has data in the cache, so that API call returns data.  Let's assume that somewhere in the time between Pusher sending a message and everything getting processed, the second Sidekiq job has completed, so that API call also returns data from cache.  The next three Sidekiq jobs have not completed, so we actually enqueue 3 Sidekiq jobs to repeat the work for those 3 bits of data.
3. Another Sidekiq jobs completes and sends a Pusher notification, so we call `getData` again.  The first and second job have data in cache, and now a third job (the one that just sent a Pusher message) has completed, so it's in cache, too.  But again some new jobs can get enqueued since they're still running.

Ultimately, I think this points to, at the very least, value in tracking Pusher notifications for each `queryString` separately and, as Emilia suggests, not re-calling `getData` until all five have returned.  There are a couple approaches to this, but I think that the easiest way to do so is to have a set of five new state values.  `preCompletedQueryReadyForRefresh` or something similar.  Crib the code from `filterMatchesHash` in an `Effect` that triggers on `pusherMessage` update to set whichever state value is appropriate.  Add a new effect that listens to all five of those states, and calls `getData` if all 5 are set.  And, per your suggestion above, just re-set those states to `false` or `null` whenever you call `getData`.  I think that covers most of the issues?

Reply:
Thanks for the thorough input, y'all! @emilia-friedberg I think that's an excellent solution. I just made this change and everything seems to be working as expected